sql: implement DESCENDING order PKs for row-level TTL

ecwall opened the pull request:
Fixes #76912

Updates the schema change validation to allow TTL tables to have DESC PK columns. The functionality is version gated behind `V23_2TTLAllowDescPK`.

Changes the TTL query bounds `WHERE` clause from this format:
```
AND (col1, col2) >= ($4, $5) AND (col1, col2) < ($2, $3)
```
to this format:
```
AND (
  (col1 > $4) OR
  (col1 = $4 AND col2 >= $5)
)
AND (
  (col1 < $2) OR
  (col1 = $2 AND col2 < $3)
)

```
which allows for DESC order keys:
```
AND (
  (col1 < $4) OR
  (col1 = $4 AND col2 <= $5)
)
AND (
  (col1 > $2) OR
  (col1 = $2 AND col2 > $3)
)
```
or mixed ASC/DESC order keys:
```
AND (
  (col1 > $4) OR
  (col1 = $4 AND col2 <= $5)
)
AND (
  (col1 < $2) OR
  (col1 = $2 AND col2 > $3)
)
```

The `ORDER BY` clause also replaces implicit ascending ordering:
```
ORDER BY col1, col2
```
with explicit ordering:
```
ORDER BY col1 ASC, col2 DESC
```

Release note (sql change): TTL supports DESC order primay key columns.

Depends on
~https://github.com/cockroachdb/cockroach/pull/102030~
~https://github.com/cockroachdb/cockroach/pull/102127~

ecwall committed:
ttl: use relation name in queries for readability

Release note: None

ecwall committed:
ttl: add selectQueryBuilder support for descending pk columns

Informs #76912

Changes the TTL query bounds `WHERE` clause from this format:
```
AND (col1, col2) >= ($4, $5) AND (col1, col2) < ($2, $3)
```
to this format:
```
AND (
  (col1 > $4) OR
  (col1 = $4 AND col2 >= $5)
)
AND (
  (col1 < $2) OR
  (col1 = $2 AND col2 < $3)
)

```
which allows for DESC order keys:
```
AND (
  (col1 < $4) OR
  (col1 = $4 AND col2 <= $5)
)
AND (
  (col1 > $2) OR
  (col1 = $2 AND col2 > $3)
)
```
or mixed ASC/DESC order keys:
```
AND (
  (col1 > $4) OR
  (col1 = $4 AND col2 <= $5)
)
AND (
  (col1 < $2) OR
  (col1 = $2 AND col2 > $3)
)
```

The `ORDER BY` clause also replaces implicit ascending ordering:
```
ORDER BY col1, col2
```
with explicit ordering:
```
ORDER BY col1 ASC, col2 DESC
```

Release note: None

ecwall committed:
ttl: merge colinfo/ttl.go into catpb/ttl.go to avoid a circular dependency

Release note: None

ecwall committed:
ttl: extract ttl expr calculation into RowLevelTTL.GetTTLExpr()

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101869)
<!-- Reviewable:end -->

ecwall committed:
ttl: fix ttljob_test subtest testing.T reference

Release note: None

ecwall committed:
sql: implement DESCENDING order PKs for row-level TTL

Fixes #76912

Updates the schema change validation to allow TTL tables to have DESC PK
columns. The functionality is version gated behind `V23_2TTLAllowDescPK`.

Release note (sql change): TTL supports DESC order primay key columns.

otan commented on pkg/sql/ttl/ttlbase/ttl_helpers.go:
```diff
@@ -12,42 +12,161 @@ package ttlbase
+			if i > 0 {
+				buf.WriteString(", ")
+			}
+			buf.WriteString(pkColNames[i])
```
i believe you need to escape these using `lex.EncodedRestrictSQLIdent`, or for extra mark, let's re-create the string using the `tree.Delete` object instead.

this seems otherwise broken when you have a pk like `"SPECIAL_name" id INT PRIMARY KEY`

ecwall replied:
This now happens here https://github.com/cockroachdb/cockroach/pull/101869/files#diff-a21f73ec9c878a961dc5a592d33403e5d2fabd925c97be208ee6b9df645b504fR114 right after getting the `pkColNames` from the descriptor since everywhere after that might need the escaped version.

otan reviewed (commented):
skimmed code, not a full review.

high level questions:
* does this well work with mixed order PKs, e.g. `PRIMARY KEY(id ASC, id2 DESC)`? not sure i see how it works with that with the optimiser liking it (this is now not simply a range scan)
* how does this go with PKs that look like `PRIMARY KEY (id ASC, id2 ASC, id3 ASC)`? what kind of queries get generated? because now you have to do `(id > 2) OR (id == 2 and id2 > 3) OR (id ==d && id2 == 3 && id3 > 4)`. not sure i see any tests for it. is the optimiser smart enough to figure out that kind of query is a range scan?

ecwall committed:
sql: name struct params during initialization so optional params can be added

Release note: None

ecwall commented:
> * does this well work with mixed order PKs, e.g. `PRIMARY KEY(id ASC, id2 DESC)`? not sure i see how it works with that with the optimiser liking it (this is now not simply a range scan)

I added `ASC -> DESC -> ASC` and `DESC -> ASC -> DESC` tests for this now. A `DESC` column just reverses the `>` and `<` for start and end bounds checks from an `ASC` column.

> * how does this go with PKs that look like `PRIMARY KEY (id ASC, id2 ASC, id3 ASC)`? what kind of queries get generated? because now you have to do `(id > 2) OR (id == 2 and id2 > 3) OR (id ==d && id2 == 3 && id3 > 4)`. not sure i see any tests for it. is the optimiser smart enough to figure out that kind of query is a range scan?
I tested it out and the `EXPLAIN` output looks similar to before. Does this represent a "range scan"? 
```
EXPLAIN (verbose) SELECT col1, col2, col3, col4                                                                                                                                                     
                        -> FROM t2                                                                                                                                                                                             
                        -> AS OF SYSTEM TIME INTERVAL '-10 seconds'                                                                                                                                                            
                        -> WHERE true                                                                                                                                                                                          
                        -> AND (                                                                                                                                                                                               
                        ->   (col1 > 1) OR                                                                                                                                                                                     
                        ->   (col1 = 1 AND col2 > 1) OR                                                                                                                                                                        
                        ->   (col1 = 1 AND col2 = 1 AND col3 > 1) OR                                                                                                                                                           
                        ->   (col1 = 1 AND col2 = 1 AND col3 = 1 AND col4 >= 1)                                                                                                                                                
                        -> )                                                                                                                                                                                                   
                        -> AND (                                                                                                                                                                                               
                        ->   (col1 < 9) OR                                                                                                                                                                                     
                        ->   (col1 = 9 AND col2 < 9) OR                                                                                                                                                                        
                        ->   (col1 = 9 AND col2 = 9 AND col3 < 9) OR                                                                                                                                                           
                        ->   (col1 = 9 AND col2 = 9 AND col3 = 9 AND col4 < 9)                                                                                                                                                 
                        -> )                                                                                                                                                                                                   
                        -> ORDER BY col1 ASC, col2 ASC, col3 ASC, col4 ASC                                                                                                                                                     
                        -> LIMIT 2;                                                                                                                                                                                            
                                     info
-------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • scan
    columns: (col1, col2, col3, col4)
    ordering: +col1,+col2,+col3,+col4
    estimated row count: 0 (11% of the table; stats collected 17 minutes ago)
    table: t2@t2_pkey
    spans: /1/1/1/1-/9/9/9/9
    limit: 2
```

ecwall commented:
Depends on https://github.com/cockroachdb/cockroach/pull/102030

rafiss reviewed (changes requested):
i just have small nits that should be easy to address. i'll mark it as "needs changes" since one that i'm worried about seems like it might be unsafe, but i'm not sure

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-:-NTtfAR17SMuYz3s1TgL:b-ppkfpz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @otan)

___
*[`pkg/sql/ttl/ttlbase/ttl_helpers.go` line 27 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTt_RrUDZw853TsKtdg:-NTt_RrUDZw853TsKtdh:b-cirb81) ([raw file](https://github.com/cockroachdb/cockroach/blob/8665b407107e0903065ef18de7ce1890a155d0a8/pkg/sql/ttl/ttlbase/ttl_helpers.go#L27)):*
> ```Go
> // TTL job.
> const SelectTemplate = `SELECT %[1]s
> FROM %[2]s
> ```

i thought we were using IDs intentionally, in case the table got renamed? or was this just tech debt?
___
*[`pkg/sql/ttl/ttlbase/ttl_test_util.go` line 15 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTtcZAT7pL9BrG2pUlh:-NTtcZAT7pL9BrG2pUli:bngvzyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/51145eff94c01d6b13848b8f8b6d29b1126ef12a/pkg/sql/ttl/ttlbase/ttl_test_util.go#L15)):*
> ```Go
> import "fmt"
> 
> // GenPKColNames generates column names col0, col1, col2, etc for tests.
> ```

can you make these look like `Col-%d`, so that we test the identifier escaping logic?
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 117 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTtconX1-WDaMF3pJ_V:-NTtconX1-WDaMF3pJ_W:bvjyyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/51145eff94c01d6b13848b8f8b6d29b1126ef12a/pkg/sql/ttl/ttljob/ttljob_processor.go#L117)):*
> ```Go
> 	for i := range pkColNames {
> 		lexbase.EncodeRestrictedSQLIdent(&buf, pkColNames[i], lexbase.EncNoFlags)
> 		pkColNames[i] = buf.String()
> ```

is it safe to modify this slice? it looks like it will directly change `primaryIndexDesc.KeyColumnNames`


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/ttl/ttlbase/ttl_helpers.go` line 27 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTt_RrUDZw853TsKtdg:-NTtyMLE5cD2MWIE62Yf:b8dr8o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8665b407107e0903065ef18de7ce1890a155d0a8/pkg/sql/ttl/ttlbase/ttl_helpers.go#L27)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i thought we were using IDs intentionally, in case the table got renamed? or was this just tech debt?
</blockquote></details>

We check that the descriptor version does not change while deleting which will cause the job to fail if the table gets renamed.
___
*[`pkg/sql/ttl/ttlbase/ttl_test_util.go` line 15 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTtcZAT7pL9BrG2pUlh:-NTumoZ9BQwIaJLQ3L6o:b-y7bnvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/51145eff94c01d6b13848b8f8b6d29b1126ef12a/pkg/sql/ttl/ttlbase/ttl_test_util.go#L15)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you make these look like `Col-%d`, so that we test the identifier escaping logic?
</blockquote></details>

`GenPKColNames` is used by lower layer tests that do not deal with the table descriptor or escaping column names. 

However, there are tests that do test quoted column names here https://github.com/cockroachdb/cockroach/blob/4fb894e0398f21a3f7d0ea9f289718ed3ac65fdd/pkg/sql/ttl/ttljob/ttljob_test.go#L669.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 117 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTtconX1-WDaMF3pJ_V:-NTtyyxt6dJNkdAgamFx:b60xi17) ([raw file](https://github.com/cockroachdb/cockroach/blob/51145eff94c01d6b13848b8f8b6d29b1126ef12a/pkg/sql/ttl/ttljob/ttljob_processor.go#L117)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is it safe to modify this slice? it looks like it will directly change `primaryIndexDesc.KeyColumnNames`
</blockquote></details>

I initialized pkColNames to a new slice and combined logic that populates the slice with logic that encodes the column names.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @otan)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 89 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTy91iQCD0Qbq8JYyF3:-NTy91iQCD0Qbq8JYyF4:bc1gyw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d31672f2da4bdd6919873516028b5e3b88a258c/pkg/sql/ttl/ttljob/ttljob_processor.go#L89)):*
> ```Go
> 		keyColumnNames := primaryIndexDesc.KeyColumnNames
> 		pkColNames = make([]string, 0, len(keyColumnNames))
> 		for _, name := range keyColumnNames {
> ```

nit: it doesn't seem necessary to define `keyColumnNames` just to use it in the following two lines and nowhere else. and it may be risky to leave a reference to the slice owned by the descriptor. but not a big deal


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101869)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 89 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/101869#-NTy91iQCD0Qbq8JYyF3:-NTyIYxRAJNRQBjKaI_E:b-365vvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d31672f2da4bdd6919873516028b5e3b88a258c/pkg/sql/ttl/ttljob/ttljob_processor.go#L89)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: it doesn't seem necessary to define `keyColumnNames` just to use it in the following two lines and nowhere else. and it may be risky to leave a reference to the slice owned by the descriptor. but not a big deal
</blockquote></details>

Ok updated


<!-- Sent from Reviewable.io -->
